File: samples/node/snapshot.test.js
const path = require('node:path');
const { BicepTestSession } = require('@anthony-c-martin/bicep-testing');

const sampleDescribe = process.env.BICEP_TEST_VALIDATE_ONLY === '1' ? describe.skip : describe;
const metadata = [
  'ddbe463a-0554-485d-b589-0b17d60cd38b',
  '28c9069e-23e8-47d2-b640-00d2e0f09616',
  'sample-rg',
  'eastus',
  'sample-deployment',
];

sampleDescribe('real-world Bicep snapshots', () => {
  let session;

  beforeAll(async () => {
    session = await BicepTestSession.create('0.43.1');
  }, 60_000);

  afterAll(() => session.dispose());

  async function snapshot(relativePath) {
    return session.snapshot(path.resolve(__dirname, '../infra', relativePath), ...metadata);
  }

  test('environment parameters select the expected topology, SKUs, and tags', async () => {
    const development = await snapshot('environment-topology/dev.bicepparam');
    const production = await snapshot('environment-topology/prod.bicepparam');

    expect(development.diagnostics).toHaveLength(0);
    expect(development.predictedResources).toEqual([
      expect.objectContaining({
        name: 'ordersdevprimary',
        sku: { name: 'Standard_LRS' },
        tags: expect.objectContaining({ environment: 'dev', dataClassification: 'internal' }),
      }),
    ]);
    expect(development.outputs.auditStorageId).toBeNull();

    expect(production.diagnostics).toHaveLength(0);
    expect(production.predictedResources.map(resource => resource.name)).toEqual([
      'ordersprodprimary',
      'ordersprodaudit',
    ]);
    expect(production.predictedResources[0]).toEqual(expect.objectContaining({
      sku: { name: 'Standard_ZRS' },
      tags: expect.objectContaining({ environment: 'prod', dataClassification: 'confidential' }),
    }));
    expect(production.predictedResources[1].sku.name).toBe('Standard_GRS');
    expect(production.outputs.auditStorageId).toContain('/storageAccounts/ordersprodaudit');
  });

  test('the security baseline catches a deliberately weakened parameter set', async () => {
    const secure = await snapshot('security-baseline/secure.bicepparam');
    const insecure = await snapshot('security-baseline/insecure.bicepparam');
    const secureStorage = secure.predictedResources.find(resource => resource.type === 'Microsoft.Storage/storageAccounts');
    const secureVault = secure.predictedResources.find(resource => resource.type === 'Microsoft.KeyVault/vaults');
    const insecureStorage = insecure.predictedResources.find(resource => resource.type === 'Microsoft.Storage/storageAccounts');

    expect(secure.diagnostics).toHaveLength(0);
    expect(secureStorage.properties).toEqual(expect.objectContaining({
      allowBlobPublicAccess: false,
      allowSharedKeyAccess: false,
      minimumTlsVersion: 'TLS1_2',
      publicNetworkAccess: 'Disabled',
      supportsHttpsTrafficOnly: true,
    }));
    expect(secureVault.properties).toEqual(expect.objectContaining({
      enablePurgeProtection: true,
      enableRbacAuthorization: true,
      publicNetworkAccess: 'Disabled',
      softDeleteRetentionInDays: 90,
    }));
    expect(insecureStorage.properties.minimumTlsVersion).toBe('TLS1_0');
    expect(insecureStorage.properties.allowBlobPublicAccess).toBe(true);
  });

  test('private endpoint, subnet, NSG, and DNS references are wired together', async () => {
    const result = await snapshot('private-network/main.bicepparam');
    const resources = Object.fromEntries(result.predictedResources.map(resource => [resource.name, resource]));

    expect(result.diagnostics).toHaveLength(0);
    expect(resources['orders-vnet'].properties.addressSpace.addressPrefixes).toEqual(['10.42.0.0/16']);
    expect(resources['orders-vnet/app'].properties).toEqual(expect.objectContaining({
      addressPrefix: '10.42.1.0/24',
      networkSecurityGroup: expect.objectContaining({ id: expect.stringContaining('/orders-app-nsg') }),
    }));
    expect(resources['orders-vnet/data'].properties).toEqual(expect.objectContaining({
      addressPrefix: '10.42.2.0/24',
      privateEndpointNetworkPolicies: 'Disabled',
      networkSecurityGroup: expect.objectContaining({ id: expect.stringContaining('/orders-data-nsg') }),
    }));
    expect(resources['orders-storage-pe'].properties.subnet.id).toBe(result.outputs.networkIds.dataSubnetId);
    expect(resources['orders-storage-pe'].properties.privateLinkServiceConnections[0].properties).toEqual(expect.objectContaining({
      groupIds: ['blob'],
      privateLinkServiceId: expect.stringContaining('/storageAccounts/ordersprivatestore'),
    }));
    expect(resources['privatelink.blob.core.windows.net/orders-vnet-link'].properties.virtualNetwork.id)
      .toBe(result.outputs.networkIds.virtualNetworkId);
  });
});

File: samples/go/snapshot_test.go
package sample_test

import (
	"context"
	"path/filepath"
	"testing"

	biceptesting "github.com/anthony-c-martin/bicep-testing/packages/go/bicep-testing"
)

func TestSnapshots(t *testing.T) {
	session := newTestSession(t)
	t.Run("environment parameters select topology, SKUs, and tags", func(t *testing.T) {
		testEnvironmentParametersSelectTopologySKUsAndTags(t, session)
	})
	t.Run("security baseline catches weakened parameters", func(t *testing.T) {
		testSecurityBaselineCatchesWeakenedParameters(t, session)
	})
	t.Run("private network references are wired together", func(t *testing.T) {
		testPrivateNetworkReferencesAreWiredTogether(t, session)
	})
}

func testEnvironmentParametersSelectTopologySKUsAndTags(t *testing.T, session *biceptesting.Session) {
	development := takeSnapshot(t, session, "environment-topology/dev.bicepparam")
	production := takeSnapshot(t, session, "environment-topology/prod.bicepparam")

	if len(development.Diagnostics) != 0 || len(development.PredictedResources) != 1 {
		t.Fatalf("development snapshot: %d diagnostics, %d resources", len(development.Diagnostics), len(development.PredictedResources))
	}
	developmentStorage := development.PredictedResources[0]
	if developmentStorage.Name != "ordersdevprimary" || nested(developmentStorage.AdditionalProperties, "sku", "name") != "Standard_LRS" {
		t.Errorf("unexpected development storage: %#v", developmentStorage)
	}
	if nested(developmentStorage.AdditionalProperties, "tags", "environment") != "dev" || development.Outputs["auditStorageId"] != nil {
		t.Error("development tags or conditional audit output did not match")
	}

	if len(production.PredictedResources) != 2 {
		t.Fatalf("production resources = %d, want 2", len(production.PredictedResources))
	}
	if production.PredictedResources[0].Name != "ordersprodprimary" || production.PredictedResources[1].Name != "ordersprodaudit" {
		t.Errorf("unexpected production topology: %#v", production.PredictedResources)
	}
	if nested(production.PredictedResources[0].AdditionalProperties, "sku", "name") != "Standard_ZRS" ||
		nested(production.PredictedResources[0].AdditionalProperties, "tags", "dataClassification") != "confidential" ||
		nested(production.PredictedResources[1].AdditionalProperties, "sku", "name") != "Standard_GRS" {
		t.Error("production SKUs or tags did not match")
	}
}

func testSecurityBaselineCatchesWeakenedParameters(t *testing.T, session *biceptesting.Session) {
	secure := takeSnapshot(t, session, "security-baseline/secure.bicepparam")
	insecure := takeSnapshot(t, session, "security-baseline/insecure.bicepparam")
	secureStorage := resourceByType(t, secure, "Microsoft.Storage/storageAccounts")
	secureVault := resourceByType(t, secure, "Microsoft.KeyVault/vaults")
	insecureStorage := resourceByType(t, insecure, "Microsoft.Storage/storageAccounts")

	if secureStorage.Properties["allowBlobPublicAccess"] != false || secureStorage.Properties["allowSharedKeyAccess"] != false ||
		secureStorage.Properties["minimumTlsVersion"] != "TLS1_2" || secureStorage.Properties["publicNetworkAccess"] != "Disabled" {
		t.Errorf("storage security baseline not applied: %#v", secureStorage.Properties)
	}
	if secureVault.Properties["enablePurgeProtection"] != true || secureVault.Properties["enableRbacAuthorization"] != true ||
		secureVault.Properties["softDeleteRetentionInDays"] != float64(90) {
		t.Errorf("vault security baseline not applied: %#v", secureVault.Properties)
	}
	if insecureStorage.Properties["minimumTlsVersion"] != "TLS1_0" || insecureStorage.Properties["allowBlobPublicAccess"] != true {
		t.Error("weakened fixture did not expose the expected regression")
	}
}

func testPrivateNetworkReferencesAreWiredTogether(t *testing.T, session *biceptesting.Session) {
	snapshot := takeSnapshot(t, session, "private-network/main.bicepparam")
	resources := map[string]biceptesting.SnapshotResource{}
	for _, resource := range snapshot.PredictedResources {
		resources[resource.Name] = resource
	}
	networkIDs := snapshot.Outputs["networkIds"].(map[string]any)
	appProperties := resources["orders-vnet/app"].Properties
	dataProperties := resources["orders-vnet/data"].Properties
	endpointProperties := resources["orders-storage-pe"].Properties
	connection := endpointProperties["privateLinkServiceConnections"].([]any)[0].(map[string]any)["properties"].(map[string]any)

	if nested(resources["orders-vnet"].Properties, "addressSpace", "addressPrefixes") == nil || appProperties["addressPrefix"] != "10.42.1.0/24" {
		t.Error("virtual network address plan did not match")
	}
	if nested(appProperties, "networkSecurityGroup", "id") == nil || dataProperties["privateEndpointNetworkPolicies"] != "Disabled" {
		t.Error("subnet security wiring did not match")
	}
	if nested(endpointProperties, "subnet", "id") != networkIDs["dataSubnetId"] || nested(connection, "privateLinkServiceId") == nil {
		t.Error("private endpoint wiring did not match outputs")
	}
	if nested(resources["privatelink.blob.core.windows.net/orders-vnet-link"].Properties, "virtualNetwork", "id") != networkIDs["virtualNetworkId"] {
		t.Error("private DNS link did not target the virtual network")
	}
}

func newTestSession(t *testing.T) *biceptesting.Session {
	t.Helper()
	session, err := biceptesting.NewSession(context.Background(), "0.46.1")
	if err != nil {
		t.Fatal(err)
	}
	t.Cleanup(func() {
		if err := session.Close(); err != nil {
			t.Errorf("close session: %v", err)
		}
	})
	return session
}

func takeSnapshot(t *testing.T, session *biceptesting.Session, relativePath string) *biceptesting.SnapshotResult {
	t.Helper()
	parametersPath, err := filepath.Abs(filepath.Join("..", "infra", relativePath))
	if err != nil {
		t.Fatal(err)
	}
	snapshot, err := session.Snapshot(context.Background(), parametersPath, biceptesting.SnapshotMetadata{
		TenantID: "ddbe463a-0554-485d-b589-0b17d60cd38b", SubscriptionID: "28c9069e-23e8-47d2-b640-00d2e0f09616",
		ResourceGroup: "sample-rg", Location: "eastus", DeploymentName: "sample-deployment",
	})
	if err != nil {
		t.Fatal(err)
	}
	return snapshot
}

func resourceByType(t *testing.T, snapshot *biceptesting.SnapshotResult, resourceType string) biceptesting.SnapshotResource {
	t.Helper()
	for _, resource := range snapshot.PredictedResources {
		if resource.Type == resourceType {
			return resource
		}
	}
	t.Fatalf("resource type %s not found", resourceType)
	return biceptesting.SnapshotResource{}
}

func nested(value map[string]any, keys ...string) any {
	var current any = value
	for _, key := range keys {
		object, ok := current.(map[string]any)
		if !ok {
			return nil
		}
		current = object[key]
	}
	return current
}
